contracts/UBI.sol

```diff
@@ -155,36 +226,34 @@ contract UBI is Initializable {
 
   /** @dev Transfers `_amount` to `_recipient` and withdraws accrued tokens.
   *  @param _recipient The entity receiving the funds.
-  *  @param _amount The amount to tranfer in base units.
+  *  @param _amount The amount to transfer in base units.
   */
   function transfer(address _recipient, uint256 _amount) public returns (bool) {
-    uint256 newSupplyFrom;
-    if (accruedSince[msg.sender] != 0 && proofOfHumanity.isRegistered(msg.sender)) {
-        newSupplyFrom = accruedPerSecond.mul(block.timestamp.sub(accruedSince[msg.sender]));
-        totalSupply = totalSupply.add(newSupplyFrom);
-        accruedSince[msg.sender] = block.timestamp;
+    uint256 _accrued;
+    if(activated[msg.sender] && proofOfHumanity.isRegistered(msg.sender)) {
+      _accrued = (block.timestamp - accruedSince[msg.sender]) * ((accruedPerSecond + stream[address(0)][msg.sender]) - stream[msg.sender][address(0)]);
+    } else if(stream[address(0)][msg.sender] != 0) {
+      _accrued = ((block.timestamp - accruedSince[msg.sender]) * stream[address(0)][msg.sender]);
     }
-    balance[msg.sender] = balance[msg.sender].add(newSupplyFrom).sub(_amount, "ERC20: transfer amount exceeds balance");
-    balance[_recipient] = balance[_recipient].add(_amount);
+    balance[msg.sender] = (balance[msg.sender] + _accrued).sub(_amount, "ERC20: transfer amount exceeds balance");
+    balance[_recipient] += _amount; // ?Public" Good for everyone if this overflows ;)
+    accruedSince[msg.sender] = block.timestamp;
     emit Transfer(msg.sender, _recipient, _amount);
     return true;
   }
   
   /** @dev Transfers `_amount` from `_sender` to `_recipient` and withdraws accrued tokens.
   *  @param _sender The entity to take the funds from.
   *  @param _recipient The entity receiving the funds.
-  *  @param _amount The amount to tranfer in base units.
+  *  @param _amount The amount to transfer in base units.
   */
   function transferFrom(address _sender, address _recipient, uint256 _amount) public returns (bool) {
-    uint256 newSupplyFrom;
-    allowance[_sender][msg.sender] = allowance[_sender][msg.sender].sub(_amount, "ERC20: transfer amount exceeds allowance");
-    if (accruedSince[_sender] != 0 && proofOfHumanity.isRegistered(_sender)) {
-        newSupplyFrom = accruedPerSecond.mul(block.timestamp.sub(accruedSince[_sender]));
-        totalSupply = totalSupply.add(newSupplyFrom);
-        accruedSince[_sender] = block.timestamp;
+    if(allowance[_sender][msg.sender] != type(uint256).max){
+      allowance[_sender][msg.sender] = allowance[_sender][msg.sender].sub(_amount, "ERC20: transfer amount exceeds allowance");      
     }
-    balance[_sender] = balance[_sender].add(newSupplyFrom).sub(_amount, "ERC20: transfer amount exceeds balance");
-    balance[_recipient] = balance[_recipient].add(_amount);       
+    balance[_sender] = (balance[_sender] + getAccruedValue(_sender)).sub(_amount, "ERC20: transfer amount exceeds balance");
```

Comment:
The call to `getAccruedValue` was inlined for performance, recommend keeping it that way.